File: dailyTemperatures.cpp
//stack
//time complexity -> O(n)
//Space complexity -> O(n)
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        std::stack<pair<int, int>> stack; //pair: {temp, index}
        vector<int> result(temperatures.size(), 0);

        for(int i = 0; i < temperatures.size(); i++){
            int temp = temperatures[i];
            while(!stack.empty() && temp > stack.top().first){
                auto pair = stack.top();
                stack.pop(); //pop the pair
                result[pair.second] = i - pair.second;
            }
            stack.push({temp, i});
        }
        return result;

    }
};

File: dailyTemperatures.py
#monotonic decreasing stack (always decreasing)
#time complexity -> O(n)
#space complexity -> O(n)
from typing import List
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack = [] #pair-list: [temp, index]
        result = [0] * len(temperatures)

        for index, temp in enumerate(temperatures):
            while stack and temp > stack[-1][0]: #stack[-1] = top, [0] means temp
                stackTemp, stackIndex = stack.pop()
                result[stackIndex] = (index - stackIndex)
            stack.append([temp, index])
        return result
    